[ShrinkWrap] Consider constant pool access as non-stack access

sushgokh opened the pull request:
As far as I understand, constant pool access does not access stack and accesses read-only memory.

This patch considers constant pool access as non-stack access allowing shrink wrapping to happen in the concerned test.

We should be seeing perf improvement with povray benchmark from SPEC17(around 12% with -flto -Ofast) after this patch.


An NFC PR #162476 already exists to upload the test before the patch but approval has got delayed. So, as @davemgreen suggested in that PR, I have uploaded the test and patch in this single PR to show how test looks like.

sushgokh committed:
[ShrinkWrap][NFC] Test with load from constant pool preventing shrink
wrapping

A future patch intends improve the situation to allow shrink wrapping happen.

sushgokh committed:
[ShrinkWrap] Consider constant pool access as non-stack access

As far as I understand, constant pool access does not access stack and
accesses read-only memory.
This patch considers constant pool access as non-stack access allowing
shrink wrapping to happen in the concerned test.
We should be seeing perf improvement with  povray benchmark from SPEC17(around 12% with -flto
-Ofast) after this patch.

MatzeB commented on llvm/test/CodeGen/AArch64/shrink-wrap-const-pool-access.mir:
```diff
@@ -0,0 +1,75 @@
+# Following code has a load from constant pool. Accessing constant pool
+# must not be considered as a stack access and hence, shrink wrapping must
+# happen.
+# CHECK:  savePoint:
```
anchor with a CHECK-LABEL?
```suggestion
# CHECK-LABEL: name: shrink_wrap_load_from_const_pool
# CHECK:  savePoint:
```

MatzeB reviewed (approved):
LGTM

fhahn reviewed (approved):
LGTM, thanks

sushgokh committed:
address comments

1. Insert CHECK-LABEL
2. Insert triple name for the test to address test failure

sushgokh committed:
Merge branch 'main' into features/sgokhale/constpool-shrinwrap
